### DISCUSSION TITLE
Restructure the package during its Swift rewrite

### DISCUSSION THREAD

**Discussion by fakeheal** [agencyenterprise/react-native-health#83] [Ideas]
Hi, it's me again!

First of all, thanks again for maintaining the repository and caring enough to migrate its codebase to Swift! ❤️ 

My "[norwich](https://i.imgur.com/nCSB5Xo.png)" in Swift is far greater than the one (almost non-existent) I have in Obj-C.

I am currently making an "explorer" for Apple Health for myself as I track so many stuff ([quanitified self](https://en.wikipedia.org/wiki/Quantified_self) all the way 🙌). I'd love to help with the rewrite, writing documentation, creating demos and testing!

Also I was wondering if structuring the code around Apple Health's documentation and "exposing" the native methods of the HealthKit library like: `HKSampleQuery`,  `HKDocumentQuery`, `HKHeartbeatSeriesQuery`, etc, would make the package easier to maintain and update (when Apple release new features).

Additionally all the "helper" methods already included in the library can be added on later stages to the code or just document how each metric can be saved/read by using the already available methods "exposed" from HealthKit to RN.

This whole post assumes I have understood what this package currently offers and your goals, I hope it makes sense 😅

**Comment by GGGava** [comment]
Hey @fakeheal 

That's great! Your help will be really really appreciated 😄 I'll soon start to create some issues so we can really begin the migration.

About your suggestion, maybe @Macelai or @lucaspbordignon can provide a better discussion, let's wait for them 😅

**Comment by lucaspbordignon** [comment]
Hey @fakeheal!

We're glad this is relevant for you and your help is more than welcome!

Speaking about the idea, I think it makes a lot of sense, as currently the methods we have doesn't follow a clear standard. If we follow Apple HealthKit's query standards, it will be easier to extend and add new functionality in the future.

My main concern is whether that would break the existing interface for our users

But overall, I think it's a more sustainable approach for the future. We could create our basis based on the three main pillars of HealthKit

1. [Queries](https://developer.apple.com/documentation/healthkit/queries) - To read data
2. [Samples](https://developer.apple.com/documentation/healthkit/samples) - To save data
3. [Data Types](https://developer.apple.com/documentation/healthkit/data_types) - To determine their types

Thoughts @GGGava @Macelai?

**Reply by fakeheal** [reply]
> My main concern is whether that would break the existing interface for our users.

I have already implemented most of the methods available in the package, so I can create a comprehensive "migration" guide, so the transition will be as painless as possible.

> as currently the methods we have doesn't follow a clear standard. 

I believe that creating a standard would make using the package and developing apps with it a lot more easier, so any developer that's currently using the package would welcome the changes. (My personal opinion, of course :D)

**Comment by lucaspbordignon** [comment]
As an idea, we could have the following structure:

- We have methods to set `HKAnchoredObjectQuery` and intercept data from HealthKit;
  -  As Apple states, *Often, it’s more efficient to set up and run a single anchored object query than to run separate sample and observer queries* (thanks @alfonsosn for the start on that side btw)
- We have methods to manually create `HKSampleQuery`
  - For historical data;
- We have methods to save data via `saveObject`;
- We have methods to properly parse inputs to the query/writer patterns;
- We have methods to manually set  `HKObserverQuery`
  - For compatibility and further flexibility;

With that, the developer would be able to assemble and use only what they really need and it is tightly coupled to the Apple HealthKit original documentation, which is better for maintenance in the long run.

Also, just as a side note, these are the initialization methods we currently have 

- `isAvailable`
- `initHealthKit` - *Could become `init`*
- `getAuthStatus` - *Could become `authorizationStatus`*

___

Would love to hear your thoughts on the discussion, so we can get started on the story writing for those 🙂 

Also, thanks @alfonsosn for getting started with the anchored queries. Feel free to contribute to the thread!

**Comment by alfonsosn** [comment]
I like this idea. I think for alpha of this swift migration, we should keep these methods queries flexible and open. I'm actually writing a tutorial series on react-native, healthkit, and swift: https://medium.com/equinox-media-tech/react-native-app-with-healthkit-swift-aa44cca12ad7

I can help by creating the anchor workout query on swift; and then, we can add a following commit with an anchor query for all other samples.
